### PR TITLE
feat: Sum and average aggregation queries

### PR DESCRIPTION
**Pull request by danieljbruce**
This PR demonstrates how we are going to leverage the current framework to support sum/average queries for later use.

### CODE REVIEW

File: src/aggregate.ts
Comment (vicb) [inline src/aggregate.ts:223]: Could you just
```js
const aggregation = this.property_ ? {property: {name: this.property_}} : {};
```

File: src/aggregate.ts
Comment (vicb) [inline src/aggregate.ts:232]: You can also drop the intermediate object and replace this line by
```js
{[this.operator]: aggregation}
```

File: src/aggregate.ts
Comment (vicb) [inline src/aggregate.ts:213]: nit: 
```js
constructor(public property_: string) {
```

and remove the fields + the assignment

File: src/aggregate.ts
Comment (danieljbruce) [reply src/aggregate.ts]: done

File: src/aggregate.ts
Comment (danieljbruce) [reply src/aggregate.ts]: done

File: src/aggregate.ts
Comment (danieljbruce) [reply src/aggregate.ts]: done

File: system-test/datastore.ts
Comment (danieljbruce) [inline system-test/datastore.ts:1414]: I can confirm that this value and the value for average are the same values received by the grpc call so the values you see here are strictly backend calculations.

File: src/request.ts
Comment (danieljbruce) [inline src/request.ts:913]: This code was taken from https://github.com/googleapis/nodejs-datastore/blob/e7c3aaf665ebae5a91fb692bb0af54717e5ea587/src/request.ts#L290-L299 and can be refactored after the merge as per the TODO.

File: src/request.ts
Comment (kolea2) [inline src/request.ts:894]: is this to be done in this PR or in a future one?

File: src/request.ts
Comment (kolea2) [reply src/request.ts:913]: any reason why we can't refactor this now?

File: system-test/datastore.ts
Comment (kolea2) [inline system-test/datastore.ts:713]: is this related to SUM/AVG changes? If not, let's move it into a separate PR

File: system-test/datastore.ts
Comment (kolea2) [inline system-test/datastore.ts:980]: might be more interesting to include a sum, average, and a count all in one query

File: system-test/datastore.ts
Comment (danieljbruce) [inline system-test/datastore.ts:1411]: Include `overflow` word here

File: src/request.ts
Comment (danieljbruce) [reply src/request.ts:913]: No reason. It is a choice between having two simpler PRs and one PR that includes both changes. Doing so now is fine though.

File: src/request.ts
Comment (danieljbruce) [reply src/request.ts:894]: Same the above comment. We can do it now if preferred.

File: system-test/datastore.ts
Comment (danieljbruce) [reply system-test/datastore.ts:713]: ok

File: system-test/datastore.ts
Comment (danieljbruce) [reply system-test/datastore.ts:973]: I could do that in a separate block. This test is inside the `with a sum filter` block.

Comment (danieljbruce) [reply system-test/datastore.ts:973]: Done

File: system-test/datastore.ts
Comment (danieljbruce) [reply system-test/datastore.ts]: Done

File: src/aggregate.ts
Comment (kolea2) [reply src/aggregate.ts:162]: this looks done, good to resolve?

File: src/aggregate.ts
Comment (kolea2) [inline src/aggregate.ts:166]: should this be updated to remove "Count" since it applies to all?

File: src/aggregate.ts
Comment (kolea2) [inline src/aggregate.ts:172]: can we add some additional testing on Count now that this is optional?

File: src/request.ts
Comment (kolea2) [inline src/request.ts:870]: what is the difference between this method and `getSharedQueryOptions`? The naming is confusing here.

File: system-test/datastore.ts
Comment (kolea2) [inline system-test/datastore.ts:680]: why are we adding this? Can you add a comment above explaining?

File: system-test/datastore.ts
Comment (kolea2) [inline system-test/datastore.ts:1258]: I would add one more test here using a combination of aliases and non aliases (aka count with an alias, sum without an alias, avg with an alias all in one query)

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: src/aggregate.ts
Comment (danieljbruce) [reply src/aggregate.ts]: Done

File: src/request.ts
Comment (danieljbruce) [reply src/request.ts:870]: `getSharedQueryOptions` also accepts a query and calls `getSharedOptionsOnly`. `getSharedOptionsOnly` is also called for fetches in `createReadStream` where options need to be built without considering a query.

File: src/aggregate.ts
Comment (danieljbruce) [reply src/aggregate.ts:172]: Yes. Added in commit called "Added tests for when an empty alias is provided".

File: system-test/datastore.ts
Comment (danieljbruce) [reply system-test/datastore.ts:1283]: Done

Comment (danieljbruce) [reply system-test/datastore.ts:1283]: https://github.com/googleapis/nodejs-datastore/pull/1097/commits/bbff19ff13c0eabb19dee64536b2e493d5982cf9

File: src/request.ts
Comment (kolea2) [reply src/request.ts:870]: Thanks for the explanation! Maybe `getSharedQueryOptions` and `getSharedRequestOptions`, or `getQueryOptions` and `getRequestOptions` would be a little more readable, wdyt?

Comment (danieljbruce) [reply src/request.ts]: `getQueryOptions` and `getRequestOptions` would work really well. The good thing about these two functions is that they are both private so we can change the names later too if need be.

File: system-test/datastore.ts
Comment (ruyadorno) [inline system-test/datastore.ts:680]: I believe it would be important to rephrase this comment so that it explains **why** we need the sleeping: 
is it simulating process time from datastore? is it simulating latency?

**Review by ruyadorno — COMMENTED** [review]
A couple of questions and tweaking a comment but otherwise it's looking good to me 👍

File: src/index.ts
Comment (danieljbruce) [reply src/index.ts:44]: See design doc comment.

File: src/index.ts
Comment (ruyadorno) [reply src/index.ts:1821]: makes sense then, so it sounds like `getSharedQueryOptions` was added here in error since it also lived in `request.ts`

Comment (danieljbruce) [reply src/index.ts:1821]: Yes. I think so.

File: system-test/datastore.ts
Comment (kolea2) [inline system-test/datastore.ts:713]: rename to "emptyData"

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: system-test/datastore.ts
Comment (danieljbruce) [reply system-test/datastore.ts]: Done

**Review by kolea2 — APPROVED** [review]
lgtm, please wait for @ruyadorno's approval as well before merging. Thanks for all the hard work on this!

**Review by ruyadorno — COMMENTED** [review]
LGTM 👍

**Pull request merged**
